atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
-                           username=self.username)
-        cmd_out = ssh_client.exec_command(cmd)
-        out = cmd_out[1].read()
-        error = cmd_out[2].read()
-        if error:
-            raise Exception('There was an error pulling the runtime: {}'.format(error.decode()))
-        ssh_client.close()
-        return out
-
-
-class ResourceManager:
-    '''
-    Manages build resources
-    '''
-
-    def __init__(self, hosts: List[Host]):
-        self.hosts = hosts
+    def _ssh_run_remote_cmd(self, client, cmd: List):
+        """
+        Run a Shell command on a remote host
+
+        :return: stdout, stderr and exitcode of shell command
+        """
 
-    def find(self) -> List[Host]:
         try:
-            resources = []
-            for host in self.hosts:
-                host.check_avail()
-                if len(host.free_slots) > 0:
-                    resources.append(host)
-            sorted_result = sorted(resources, key=lambda h: len(h.free_slots))
-            logger.debug('sorted result: %s', sorted_result)
-            return sorted_result
-        except Exception as err:
-            logger.debug(err)
-            logger.error("Unable to read hosts file")
-            return []
-
-    def obtain_free_resource(self, prid: str) -> Tuple[Host, int]:
-        availiable_hosts = self.find()
-        for host in availiable_hosts:
-            for sid in host.free_slots:
-                if host.lock(sid, prid):
-                    return host, sid
-        raise Exception('No availiable host found')
-
-
-if __name__ == '__main__':
-    my_hosts = [Host('username', 'hostname')]
-    mngr = ResourceManager(my_hosts)
-    # locking
-    free_host, slot = mngr.obtain_free_resource('unique-name-for-pipeline-run')
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+            stdin, stdout, stderr = client.exec_command(
+                cmd, timeout=SSH_COMMAND_TIMEOUT)
+        except paramiko.ssh_exception.AuthenticationException as ex:
+            logger.warning("SSH authentication failed. %s", ex) #  TODO: Question - Probably not needed for exec command
+        except paramiko.ssh_exception.ChannelException as ex:
+            logger.warning("Attempt to open a new Channel failed. %s", ex) #  TODO: Question - Probably not needed for exec command
+        except paramiko.ssh_exception.NoValidConnectionsError as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        except paramiko.ssh_exception.SSHException as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        return stdout, stderr, client.recv_exit_status()
+
+    def _ssh_run_remote_cmd_to_get_json(self, client, cmd: List):
+        stdin, stdout, stderr = self._ssh_run_remote_cmd(client, cmd)
+        try:
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON  #  TODO: Question - Do we want to fail build on loading invalid json from remote host????
```

Comment:
Skip, don't fail build. We don't want to explode all builds because one slot/host is misconfigured